### PR TITLE
Clearer classic mode label equivalence checks

### PR DESCRIPTION
**Pull request by vivianyyd**
This PR is ready for review; the diffs are inflated due to containing changes in #1661 and #1671 but should go away once they are merged. The relevant commit to review is b1399b6, parent b020ff3.

There were some calls to `is_optional` used to check whether labels are equivalent in classic mode, where argument labels are ignored. This replaces those calls with a clearer check.

### CODE REVIEW

File: ocaml/typing/ctype.ml [hunk 1]
```diff
@@ -2677,6 +2677,11 @@ let rec expands_to_datatype env ty =
       end
   | _ -> false
 
+let equivalent_in_classic_mode l1 l2 =
+  l1 = l2 || (match l1, l2 with
+  | (Nolabel | Labelled _), (Nolabel | Labelled _) -> true
```
Comment (ncik-roberts) [inline ocaml/typing/ctype.ml:2682]: Two `Labelled` args with different labels are also equivalent in classic mode.

Comment (vivianyyd) [reply ocaml/typing/ctype.ml:2682]: Yes, that's covered by the first case in the `match`

Comment (ncik-roberts) [reply ocaml/typing/ctype.ml:2682]: And you are very right (thanks). I should remember the difference between commas and vertical bars.

File: ocaml/typing/ctype.ml [hunk 2]
```diff
@@ -3844,7 +3849,7 @@ let filter_arrow env t l ~force_tpoly =
       constrain_type_layout_exn env Unify t' layout;
       arrow_desc
   | Tarrow((l', arg_mode, ret_mode), ty_arg, ty_ret, _) ->
-      if l = l' || !Clflags.classic && l = Nolabel && not (is_optional l')
+      if l = l' || !Clflags.classic && l = Nolabel && not (is_omittable l')
```
Comment (goldfirere) [inline ocaml/typing/ctype.ml:3852]: I would still use `equivalent_in_classic_mode` here, not `is_omittable`. Yes, it's a little duplicative, but it's clearer to readers.

File: ocaml/typing/ctype.ml [hunk 3]
```diff
@@ -2677,6 +2677,11 @@ let rec expands_to_datatype env ty =
       end
   | _ -> false
 
+let equivalent_in_classic_mode l1 l2 =
```
Comment (goldfirere) [inline ocaml/typing/ctype.ml:2680]: I know the check is for `Clflags.classic`, but I find that name very strange. Let's rename this to `equivalent_with_nolabels`.

File: ocaml/typing/ctype.ml (hunk 2 above)
Comment (vivianyyd) [reply ocaml/typing/ctype.ml:3852]: This accepts a strict subset of cases accepted by `equivalent_in_classic_mode`; if `l=Labelled x` and `l'=Nolabel y` then the check fails (or if they're both labelled, but with different names). I did want to ask why this case is different from the others though. 
Intuitively, I see that `l'` is the expected label, and trying to match that up with a function that doesn't satisfy either A. same label or B. is unlabelled when the expected is labelled doesn't quite work. But I don't completely understand why we care about this restriction here when we don't in other places.

Comment (goldfirere) [reply ocaml/typing/ctype.ml:3852]: Oh -- I was thinking to have `l = Nolabel && equivalent_in_classic_mode l l'`, which should accept the same situations

**Review by goldfirere — APPROVED** [review]

**Pull request merged**